qubesadmin/tools/qvm_template.py

```diff
@@ -0,0 +1,1546 @@
+            if weight in weight_types:
+                fields.append(field)
+        exact = all(x[-1] for x in needles)
+        match = 'Exactly Matched' if exact else 'Matched'
+        keywords = sorted(list(set(x[1] for x in needles)))
+        return ' & '.join(fields) + ' ' + match + ': ' + ', '.join(keywords)
+
+    last_header = ''
+    for idx, needles in search_res:
+        # Print headers
+        cur_header = gen_header(needles)
+        if last_header != cur_header:
+            last_header = cur_header
+            # XXX: The style is different from that of DNF
+            print('===', cur_header, '===')
+        print(query_res[idx].name, ':', query_res[idx].summary)
+
+def remove(
+        args: argparse.Namespace,
+        app: qubesadmin.app.QubesBase,
+        disassoc: bool = False,
+        purge: bool = False,
+        dummy: str = 'dummy'
+        ) -> None:
+    """Command that remove templates.
+
+    :param args: Arguments received by the application.
+    :param app: Qubes application object
+    :param disassoc: Whether to disassociate VMs from the templates
+    :param purge: Whether to remove VMs based on the templates
+    :param dummy: Name of dummy VM if disassoc is used
+    """
+    # NOTE: While QubesArgumentParser provide similar functionality
+    #       it does not seem to work as a parent parser
+    for tpl in args.templates:
+        if tpl not in app.domains:
+            parser.error("no such domain: '%s'" % tpl)
+
+    remove_list = args.templates
+    if purge:
+        # Not disassociating first may result in dependency ordering issues
+        disassoc = True
+        # Remove recursively via BFS
+        remove_set = set(remove_list) # visited
+        idx = 0
+        while idx < len(remove_list):
+            tpl = remove_list[idx]
+            idx += 1
+            vm = app.domains[tpl]
+            for holder, prop in qubesadmin.utils.vm_dependencies(app, vm):
+                if holder is not None and holder.name not in remove_set:
+                    remove_list.append(holder.name)
+                    remove_set.add(holder.name)
+
+    if not args.yes:
+        repeat = 3 if purge else 1
+        # XXX: Mutating the list later seems to break the tests...
+        remove_list_copy = remove_list.copy()
+        for _ in range(repeat):
+            confirm_action(
+                'This will completely remove the selected VM(s)...',
+                remove_list_copy)
+
+    if disassoc:
+        # Remove the dummy afterwards if we're purging
+        # as nothing should depend on it in the end
+        remove_dummy = purge
+        # Create dummy template; handle name collisions
+        orig_dummy = dummy
+        cnt = 1
+        while dummy in app.domains \
+                and app.domains[dummy].features.get(
+                    'template-dummy', '0') != '1':
+            dummy = '%s-%d' % (orig_dummy, cnt)
+            cnt += 1
+        if dummy not in app.domains:
```

Comment:
Is there a preferred way to mock constructs like this?

Existing tests that create VMs seem to just put the created VM in `admin.vm.List`, which interferes with the `if` here.